src/validation.cpp

```diff
@@ -2468,6 +2482,18 @@ bool CChainState::DisconnectTip(BlockValidationState& state, DisconnectedBlockTr
         assert(flushed);
     }
     LogPrint(BCLog::BENCH, "- Disconnect block: %.2fms\n", (GetTimeMicros() - nStart) * MILLI);
+
+    {
+        // Prune locks that began at or after the tip should get moved backward so they get a chance to reorg
```

Comment:
In commit "blockstorage: Add prune locks to BlockManager" (d249636878b39aa34222d335cd85f0e2561bd50a)

Note just for understanding this: Interesting to see this logic added, since it wasn't in original version of the PR. It looks like the result of discussion with Luke in https://github.com/bitcoin/bitcoin/pull/21726#discussion_r785479282 and is it checked by a python test looking for the "lock moved back to" log message below.

IMO, this change adds a little bit of internal complexity but improves the external design and reliability of prune locks. If you think about indexes running asynchronously behind the main chain, it's a nice thing if prune locks are updated automatically when there's a reorg, and indexes are free to catch up to block connected and disconnected notifications at any time later. They don't need to suddenly hurry and update their locks because this code does it for them.